src/picongpu/include/fields/background/templates/TWTS/BField.tpp

```diff
@@ -0,0 +1,409 @@
+#include "fields/background/templates/TWTS/GetInitialTimeDelay_SI.tpp"
+#include "fields/background/templates/TWTS/getFieldPositions_SI.tpp"
+#include "fields/background/templates/TWTS/BField.hpp"
+
+namespace picongpu
+{
+/** Load pre-defined background field */
+namespace templates
+{
+/** Traveling-wave Thomson scattering laser pulse */
+namespace twts
+{
+    namespace pmMath = PMacc::algorithms::math;
+    
+    HINLINE
+    BField::BField( const float_64 focus_y_SI,
+                    const float_64 wavelength_SI,
+                    const float_64 pulselength_SI,
+                    const float_64 w_x_SI,
+                    const float_64 w_y_SI,
+                    const float_X phi,
+                    const float_X beta_0,
+                    const float_64 tdelay_user_SI,
+                    const bool auto_tdelay ) :
+        focus_y_SI(focus_y_SI), wavelength_SI(wavelength_SI),
+        pulselength_SI(pulselength_SI), w_x_SI(w_x_SI),
+        w_y_SI(w_y_SI), phi(phi), beta_0(beta_0),
+        tdelay_user_SI(tdelay_user_SI), dt(SI::DELTA_T_SI),
+        unit_length(UNIT_LENGTH), auto_tdelay(auto_tdelay)
+    {
+        /* Note: Enviroment-objects cannot be instantiated on CUDA GPU device. Since this is done
+         * on host (see fieldBackground.param), this is no problem. */
+        const SubGrid<simDim>& subGrid = Environment<simDim>::get().SubGrid();
+        halfSimSize = subGrid.getGlobalDomain().size / 2;
+        tdelay = detail::getInitialTimeDelay_SI(auto_tdelay, tdelay_user_SI, 
+                                                halfSimSize, pulselength_SI,
+                                                focus_y_SI, phi, beta_0);
+    }
+    
+    template<>
+    HDINLINE float3_X
+    BField::getTWTSBfield_Normalized<DIM3>(
+            const PMacc::math::Vector<floatD_64,detail::numComponents>& bFieldPositions_SI,
+            const float_64 time) const
+    {
+        PMacc::math::Vector<float3_64,detail::numComponents> pos(0.0);
+        for (uint32_t k = 0; k<detail::numComponents;++k) {
+            for (uint32_t i = 0; i<simDim;++i) pos[k][i] = bFieldPositions_SI[k][i];
+        }
+        
+        /* Calculate By-component with the Yee-Cell offset of a By-field */
+        const float_64 By_By = calcTWTSBy(pos[1], time);
+        /* Calculate Bz-component the Yee-Cell offset of a By-field */
+        const float_64 Bz_By = calcTWTSBz(pos[1], time);
+        /* Calculate By-component the Yee-Cell offset of a Bz-field */
+        const float_64 By_Bz = calcTWTSBy(pos[2], time);
+        /* Calculate Bz-component the Yee-Cell offset of a Bz-field */
+        const float_64 Bz_Bz = calcTWTSBz(pos[2], time);
+        /* Since we rotated all position vectors before calling calcTWTSBy and calcTWTSBz,
+         * we need to back-rotate the resulting B-field vector. */
+        /* RotationMatrix[-(PI/2+phi)].(By,Bz) for rotating back the Field-Vektors. */
```

Comment:
@ax3l Note that this is currently unique code that applies to the B-field only (because of intra-cell position offsets). My intention is to generalize this in the upcoming feature update when I introduce different polarizations for the field.